rln/src/merkle.rs

```diff
@@ -1,26 +1,36 @@
 // Adapted from https://github.com/kilic/rln/blob/master/src/merkle.rs
 //
 use crate::poseidon::{Poseidon as Hasher, PoseidonParams};
-// TODO Replace these with arkworks
-use sapling_crypto::bellman::pairing::ff::{Field, PrimeField, PrimeFieldRepr};
-use sapling_crypto::bellman::pairing::Engine;
+use ark_ff::{Field, PrimeField};
+//use ff::{Field, PrimeField, PrimeFieldRepr};
+use ark_ec::PairingEngine;
+
 use std::io::{self, Error, ErrorKind};
 use std::{collections::HashMap, hash::Hash};
+use std::str::FromStr;
+use std::fmt::Debug;
+
+use ark_bn254::Fr;
 
 pub struct IncrementalMerkleTree<E>
 where
-    E: Engine,
+    E: PairingEngine,
 {
     pub current_index: usize,
     merkle_tree: MerkleTree<E>,
 }
 
+// XXX: I don't understand this second where clause
 impl<E> IncrementalMerkleTree<E>
 where
-    E: Engine,
+    E: PairingEngine,
+    <<E as PairingEngine>::Fr as FromStr>::Err: Debug,
 {
+    // XXX Some kind of trait bound thing?
     pub fn empty(hasher: Hasher<E>, depth: usize) -> Self {
         let mut zero: Vec<E::Fr> = Vec::with_capacity(depth + 1);
+        // XXX I don't understand why I can't call from_str here
+        // XXX minimal example?
         zero.push(E::Fr::from_str("0").unwrap());
```

Comment:
you need to specify that you want to use `Fr` as a `FromStr`. Something like `E::<Fr as FromStr>::from_str` might work.